storage/src/adb/keyless.rs

```diff
@@ -0,0 +1,863 @@
+    /// does not correspond to an append.
+    pub async fn get(&self, loc: u64) -> Result<Option<V>, Error> {
+        match self.locations.read(loc).await {
+            Ok(op) => {
+                if let Operation::Append(offset) = op {
+                    let section = loc / self.values_per_section;
+                    let Some(v) = self.values.get(section, offset).await? else {
+                        panic!("didn't find value at location {loc} and offset {offset}");
+                    };
+                    Ok(Some(v))
+                } else {
+                    Ok(None)
+                }
+            }
+            Err(e) => Err(Error::Journal(e)),
+        }
+    }
+
+    /// Get the number of appends + commits that have been applied to the db.
+    pub fn size(&self) -> u64 {
+        self.size
+    }
+
+    /// Returns the section of the values journal where we are currently writing new items.
+    fn current_section(&self) -> u64 {
+        self.size / self.values_per_section
+    }
+
+    /// Return the oldest location that remains retrievable.
+    pub async fn oldest_retained_loc(&self) -> Result<Option<u64>, Error> {
+        self.locations
+            .oldest_retained_pos()
+            .await
+            .map_err(Error::Journal)
+    }
+
+    /// Prunes the db of up to all operations that have location less than `loc`. The actual number
+    /// pruned may be fewer than requested due to blob boundaries in the underlying journals.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `loc` is greater than the current size of the database.
+    pub async fn prune(&mut self, loc: u64) -> Result<(), Error> {
+        assert!(loc <= self.size);
+
+        // Prune the locations journal first. This ensures that if pruning fails, we never have
+        // location entries without corresponding data in the other structures.
+        self.locations.prune(loc).await?;
+
+        // Prune the MMR and values journals to the corresponding positions.
+        let prune_to_section = loc / self.values_per_section;
+        try_join!(
+            self.mmr
+                .prune_to_pos(&mut self.hasher, leaf_num_to_pos(loc))
+                .map_err(Error::Mmr),
+            self.values.prune(prune_to_section).map_err(Error::Journal),
+        )?;
+
+        Ok(())
+    }
+
+    /// Append a value to the db, returning its location which can be used to retrieve it.
+    pub async fn append(&mut self, value: V) -> Result<u64, Error> {
+        let loc = self.size;
+        let section = self.current_section();
+        let (offset, _) = self.values.append(section, value).await?;
+        self.locations.append(Operation::Append(offset)).await?;
+        self.mmr
+            .add_batched(&mut self.hasher, &Operation::Append(offset).encode())
+            .await?;
+
+        self.size += 1;
+        if section != self.current_section() {
+            self.values.sync(section).await?;
```

Comment:
Should we also sync `self.locations` here (I think the current recovery pattern doesn't require it but figure I'd call out in case)?

Reply:
Actually, I think `journal::Fixed` will call sync automagically behind the scenes if configured to have the same `items_per_section`